src/command_func.h

```diff
@@ -64,60 +65,417 @@ static inline DoCommandFlag CommandFlagsToDCFlags(CommandFlags cmd_flags)
+#	pragma GCC diagnostic push
+#	pragma GCC diagnostic ignored "-Wcast-function-type"
+#	define SILENCE_GCC_FUNCTION_POINTER_CAST
+#endif
+
+template<Commands TCmd, typename T, bool THasTile> struct DoCommandHelper;
+
+class DoCommandHelperBase {
+protected:
+	static void InternalDoBefore(bool top_level, bool test);
+	static void InternalDoAfter(CommandCost &res, DoCommandFlag flags, bool top_level, bool test);
+	static std::tuple<bool, bool, bool> InternalPostBefore(Commands cmd, CommandFlags flags, TileIndex tile, StringID err_message, bool network_command);
+	static void InternalPostResult(const CommandCost &res, TileIndex tile, bool estimate_only, bool only_sending, StringID err_message, bool my_cmd);
+	static bool InternalExecutePrepTest(CommandFlags cmd_flags, TileIndex tile, Backup<CompanyID> &cur_company);
+	static std::tuple<bool, bool, bool> InternalExecuteValidateTestAndPrepExec(CommandCost &res, CommandFlags cmd_flags, bool estimate_only, bool network_command, Backup<CompanyID> &cur_company);
+	static CommandCost InternalExecuteProcessResult(Commands cmd, CommandFlags cmd_flags, const CommandCost &res_test, const CommandCost &res_exec, Money extra_cash, TileIndex tile, Backup<CompanyID> &cur_company);
+	static void LogCommandExecution(Commands cmd, StringID err_message, TileIndex tile, const CommandDataBuffer &args, bool failed);
+};
+
+/**
+ * Templated wrapper that exposes the command parameter arguments
+ * on the various DoCommand calls.
+ * @tparam Tcmd The command-id to execute.
+ * @tparam Tret Return type of the command.
+ * @tparam Targs The command parameter types.
+ */
+template <Commands Tcmd, typename Tret, typename... Targs>
+struct DoCommandHelper<Tcmd, Tret(*)(DoCommandFlag, Targs...), true> : protected DoCommandHelperBase {
+private:
+	/** Extract the \c CommandCost from a command proc result. */
+	static inline CommandCost &ExtractCommandCost(Tret &ret)
+	{
+		if constexpr (std::is_same_v<Tret, CommandCost>) {
+			return ret;
+		} else {
+			return std::get<0>(ret);
+		}
+	}
+
+	/** Make a command proc result from a \c CommandCost. */
+	static inline Tret MakeResult(const CommandCost &cost)
+	{
+		Tret ret{};
+		ExtractCommandCost(ret) = cost;
+		return ret;
+	}
+
+public:
+	/**
+	 * This function executes a given command with the parameters from the #CommandProc parameter list.
+	 * Depending on the flags parameter it execute or test a command.
+	 *
+	 * @param flags Flags for the command and how to execute the command
+	 * @param args Parameters for the command
+	 * @see CommandProc
+	 * @return the cost
+	 */
+	static Tret Do(DoCommandFlag flags, Targs... args)
+	{
+		if constexpr (std::is_same_v<TileIndex, std::tuple_element_t<0, std::tuple<Targs...>>>) {
+			/* Do not even think about executing out-of-bounds tile-commands. */
+			TileIndex tile = std::get<0>(std::make_tuple(args...));
+			if (tile != 0 && (tile >= MapSize() || (!IsValidTile(tile) && (flags & DC_ALL_TILES) == 0))) return MakeResult(CMD_ERROR);
```

Comment:
I `CmdLevelLand` there is a check for a tile in formerly in p1 where it returns `{ CMD_ERROR, 0, INVALID_TILE }`. This would be returning `{ CMD_ERROR, 0, 0 }`? Is that going to be a problem?

Reply:
Well, it would cause a wrong red error highlight square, but I think it is safe to assume that any invalid tile would either be a straight up bug in OTTD or somebody sending faked commands over the network. And in that case I don't think the red error square is that important.